src/doc/rgw/multisite-reshard.md

```diff
@@ -0,0 +1,79 @@
+
+An additional goal is to move replication logs out of omap (so out of the bucket index) into separate rados objects. To enable this, the bucket instance metadata should be able to describe a bucket whose *index layout* is different from its *log layout*. For existing buckets, the two layouts would be identical and share the bucket index objects. Alternate log layouts are otherwise out of scope for this design.
+
+To support peer zones that are still processing old logs, the local bucket instance metadata must track the history of all log layouts that haven't been fully trimmed yet. Once bilog trimming advances past an old generation, it can delete the associated rados objects and remove that layout from the bucket instance metadata. To prevent this history from growing too large, we can refuse to reshard bucket index logs until trimming catches up.
+
+The distinction between *index layout* and *log layout* is important, because incremental sync only cares about changes to the *log layout*. Changes to the *index layout* would only effect full sync, which uses a custom RGWListBucket extension to list the objects of each index shard separately. But by changing the scope of full sync from per-bucket-shard to per-bucket and using a normal bucket listing to get all objects, we can make full sync independent of the *index layout*. And once the replication logs are moved out of the bucket index, dynamic resharding is free to change the *index layout* as much as it wants with no effect on multisite replication.
+
+## Tasks
+
+### Bucket Reshard
+
+* Modify existing state machine for bucket reshard to mutate its existing bucket instance instead of creating a new one.
+
+* Add fields for log layout. When resharding a bucket whose logs are in the index:
+    - Add a new log layout generation to the bucket instance
+    - Copy the bucket index entries into their new index layout
+    - Commit the log generation change so new entries will be written there
+    - Create a datalog entry with the new log generation
+
+### Metadata Sync
+
+* When sync fetches a bucket instance from the master zone, preserve any private fields in the local instance. Use cls_version to guarantee that we write back the most recent version of those private fields.
+
+### Data Sync
+
+* Datalog entries currently include a bucket shard number. We need to add the log generation number to these entries so we can tell which sharding layout it refers to. If we see a new generation number, that entry also implies an obligation to finish syncing all shards of prior generations.
+
+### Bucket Sync Status
+
+* Move bucket sync status to a per-bucket object, rather than having separate status per-shard.
+    - Track current log generation and an array of markers for each shard in that generation.
```

Comment:
See my comments later re bucket sync. In general I'm not sure that keeping everything in a single object makes life better. Separating it to object per shard makes life easier wrt concurrent updates, and it makes sure this can't blow up if we have _a lot_ of shards.

Reply:
that makes sense, yeah. i think the only requirement is that we have a single place to track bucket full sync, since it's no longer per-shard. but the incremental status could still be split by shard

Reply:
on second thought, i think we're adding enough complexity at the bucket level to track incremental sync across generations that it will be easier to manage that per-shard state in one place. mainly because the per-bucket sync status will track the current generation, and we'll need an easy way to tell once all of its shards have completed sync of that generation

when transitioning between generations, it's also easier to resize the centralized vector of shard states than it is to create new shard objects and garbage-collect old ones

> blow up if we have _a lot_ of shards

in https://github.com/ceph/ceph/pull/30795, we added a rgw_max_dynamic_shards=1999. i think that default is sufficient to keep this state from growing unreasonably large